src/DataLoader.Uno/DataLoaderViewController.cs

```diff
@@ -0,0 +1,407 @@
+
+			_delegate.SetState(viewState);
+
+			if (view != null)
+			{
+				var dataVisualState = GetDataVisualState(safeState);
+				GoToState(view, DataVisualGroup, dataVisualState);
+
+				var errorVisualState = GetErrorVisualState(safeState);
+				GoToState(view, ErrorVisualGroup, errorVisualState);
+
+				var loadingVisualState = GetLoadingStateVisualState(safeState);
+				GoToState(view, LoadingVisualGroup, loadingVisualState);
+
+				var combinedVisualState = $"{dataVisualState}_{errorVisualState}_{loadingVisualState}";
+				GoToState(view, CombinedVisualGroup, combinedVisualState);
+
+				if (_isLoaded)
+				{
+					_isVisualStateRefreshRequired = false;
+
+					if (_logger.IsEnabled(LogLevel.Information))
+					{
+						_logger.LogInformation($"Updated VisualStates to '{combinedVisualState}'.");
+					}
+				}
+				else
+				{
+					// If the control isn't loaded, the visual states are not applied.
+					_isVisualStateRefreshRequired = true;
+
+					if (_logger.IsEnabled(LogLevel.Debug))
+					{
+						_logger.LogTrace($"Saved VisualStates to '{combinedVisualState}'. (This control isn't loaded yet.)");
+					}
+				}
+			}
+
+			_lastUpdate = DateTime.Now;
+			_lastState = safeState;
+			_lastViewState = viewState;
+		}
+
+		/// <summary>
+		/// Applies the already saved visual states.
+		/// </summary>
+		private void RefreshVisualStates()
+		{
+			var view = _delegate.GetView();
+			if (view != null)
+			{
+				RefreshVisualState(view, DataVisualGroup);
+				RefreshVisualState(view, ErrorVisualGroup);
+				RefreshVisualState(view, LoadingVisualGroup);
+				var combinedVisualState = RefreshVisualState(view, CombinedVisualGroup);
+
+				_isVisualStateRefreshRequired = false;
+
+				if (_logger.IsEnabled(LogLevel.Information))
+				{
+					_logger.LogInformation($"Updated VisualStates to '{combinedVisualState}'.");
+				}
+			}
+		}
+
+		private static string GetDataVisualState(IDataLoaderState state)
+		{
+			if (state.IsInitial)
+			{
+				return InitialVisualStateName;
+			}
+			else if (state.IsEmpty)
+			{
+				return EmptyVisualStateName;
+			}
+			else
+			{
+				return DataVisualStateName;
+			}
+		}
+
+		private static string GetLoadingStateVisualState(IDataLoaderState state)
+		{
+			return state.IsLoading
+				? LoadingVisualStateName
+				: NotLoadingVisualStateName;
+		}
+
+		private static string GetErrorVisualState(IDataLoaderState state)
+		{
+			return state.Error != null
+				? ErrorVisualStateName
+				: NoErrorVisualStateName;
+		}
+
+		private DataLoaderViewState CreateDataLoaderViewState(IDataLoaderState state, Control view)
+		{
+			return new DataLoaderViewState()
+			{
+				View = view,
+				Parent = _delegate.GetDataContext(),
+				Source = _dataLoader,
+				Request = state.Request,
+				Data = state.Data,
+				Error = state.Error,
+				IsLoading = state.IsLoading,
+				IsInitial = state.IsInitial,
+				IsEmpty = state.IsEmpty
+			};
+		}
+
+		private async Task RunOnDispatcher(Action action)
```

Comment:
Maybe similar question to LockAsync, shouldn't we probably move this one away from the control at least? Like an extension method, unless we could reuse something else